swift/test/ios/TestDriverApp/ControllerI.swift

```diff
@@ -207,45 +230,104 @@ class ControllerHelperI: ControllerHelper, TextWriter, @unchecked Sendable {
     }
 
     public func waitReady(timeout: Int32) async throws {
-        let deadline = ContinuousClock.now + .seconds(Int(timeout))
+        enum ReadyResult {
+            case completed(Int32)
+            case ready
+            case timeout
+        }
 
-        while ContinuousClock.now < deadline {
-            // Check current state
-            let state = _lock.withLock { $0 }
+        let result = await withTaskGroup(of: ReadyResult.self) { group in
+            group.addTask {
+                try? await Task.sleep(for: .seconds(Int(timeout)))
+                return .timeout
+            }
 
-            if state.isReady {
-                return
+            group.addTask {
+                for await _ in self._readyStream {
+                    break
+                }
+                return .ready
             }
 
-            if let status = state.completedStatus {
-                if status == 0 {
-                    return
-                } else {
-                    throw CommonProcessFailedException(reason: self._out)
+            group.addTask {
+                // The task cancellation from another thread that's also reading the stream
+                // can cause this task stream's iterator to return. We just loop until
+                // we get the completed status or the task is cancelled.
+                var completedStatus: Int32?
+                while completedStatus == nil && !Task.isCancelled {
```

Comment:
Why is the loop required? What about something like:

```swift
var iterator = self._completedStream.makeAsyncIterator()
_ = await iterator.next()

if Task.isCancelled {
    return .timeout
} else {
    let status = self._state.withLock { $0.completedStatus! }
    return .completed(status)
}
```

Reply:
What seems to happen here is that different iterators return the same next task, when we cancel this task if affects all callers event if using different iterators because there is a single shared task.